pkg/controllers/vdb/sandboxsubcluster_reconciler.go

```diff
@@ -153,27 +158,86 @@ func (s *SandboxSubclusterReconciler) sandboxSubclusters(ctx context.Context) (c
-				return errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+
+	// We can simply loop over the scSbMap and sandbox each subcluster. However,
+	// we want to sandbox in a determinstic order because the first subcluster
+	// in a sandbox is the primary.
+	for i := range s.Vdb.Spec.Sandboxes {
+		vdbSb := &s.Vdb.Spec.Sandboxes[i]
+		for j := range vdbSb.Subclusters {
+			sc := vdbSb.Subclusters[j].Name
+			sb, found := scSbMap[sc]
+			if !found {
+				// assume it is already sandboxed
+				continue
 			}
+			// The first subcluster in a sandbox turns into a primary. Set
+			// state in the vdb to indicate that.
+			if j == 0 {
+				_, err := vk8s.UpdateVDBWithRetry(ctx, s.VRec, s.Vdb, func() (bool, error) {
+					scMap := s.Vdb.GenSubclusterMap()
+					vdbSc, found := scMap[sc]
+					if !found {
+						return false, fmt.Errorf("subcluster %q missing in vdb %q", sc, s.Vdb.Name)
+					}
+					vdbSc.Type = vapi.SandboxPrimarySubcluster
+					return true, nil
+				})
+				if err != nil {
+					return ctrl.Result{}, errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+				}
+			}
+			res, err := s.sandboxSubcluster(ctx, sc, sb)
+			if verrors.IsReconcileAborted(res, err) {
+				// when one subcluster failed to be sandboxed, update sandbox status and return error
+				return res, errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+			}
+			succeedSbScMap[sb] = append(succeedSbScMap[sb], sc)
+			// create/update a sandbox config map
+			if _, ok := seenSandboxes[sb]; !ok {
+				err = s.checkSandboxConfigMap(ctx, sb)
+				if err != nil {
+					// when creating/updating sandbox config map failed, update sandbox status and return error
+					return ctrl.Result{}, errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+				}
+			}
+			seenSandboxes[sb] = struct{}{}
 		}
-		seenSandboxes[sb] = struct{}{}
 	}
-	return s.updateSandboxStatus(ctx, succeedSbScMap)
+	return ctrl.Result{}, s.updateSandboxStatus(ctx, succeedSbScMap)
+}
+
+// findInitiatorIPs returns the IPs to pass down to vclusterops as the initiator
+// nodes. The number of IPs could be one or two. If two, they are separated with
+// a comma.
+func (s *SandboxSubclusterReconciler) findInitiatorIPs(ctx context.Context, sandbox string) ([]string, ctrl.Result, error) {
+	// We already have an IP of an up pod from the main cluster. If we are
+	// adding to an existing sandbox, then we need to find an IP from a pod in
+	// that sandbox too.
+	_, found := s.InitiatorIPs[sandbox]
+	if !found {
+		pfs := s.PFacts.Copy(sandbox)
+		if err := pfs.Collect(ctx, s.Vdb); err != nil {
+			return nil, ctrl.Result{}, err
+		}
+		// If this is the first pod in the sandbox, then only an API from the
+		// main cluster is needed.
+		if len(pfs.Detail) == 0 {
+			return []string{s.InitiatorIPs[vapi.MainCluster]}, ctrl.Result{}, nil
+		}
+		pf, found := pfs.findFirstPodSorted(func(v *PodFact) bool {
+			return v.upNode
```

Comment:
Use primary up node should be better 
```suggestion
			return v.upNode && v.IsSandboxPrimary()
```

Reply:
Yeah, I see your point. We should be using a primary. I think you mean `v.upNode && v.IsPrimary()` though?